src/psbtv2.ts

```diff
@@ -0,0 +1,910 @@
+function bufferize (psbt:string|Buffer):Buffer {
+  let buf:Buffer|undefined;
+
+  if (Buffer.isBuffer(psbt)) {
+    buf = psbt;
+  } else if (typeof psbt === 'string' && (/^([0-9A-Fa-f][0-9A-Fa-f])+$/).test(psbt)) {
+    // Is a hex string
+    buf = Buffer.from(psbt, 'hex');
+  } else if (
+    typeof psbt === 'string' && 
+    (/^(?:[A-Za-z0-9+/]{4})*(?:[A-Za-z0-9+/]{2}==|[A-Za-z0-9+/]{3}=|[A-Za-z0-9+/]{4})$/).test(psbt)
+  ) {
+    // Is a base64 string
+    buf = Buffer.from(psbt, 'base64');
+  } else {
+    throw ERROR_PSBT_NOT_VALID;
+  }
+
+  return buf;
+}
+
+
+// Some keytypes have keydata which allows for multiple unique keys of the same
+// keytype. Getters which return values from these keys should search and return
+// values from all keys of that keytype. This function matches on the first byte
+// of each key string (hex encoded) and returns all values associated with those
+// keys as an array of string (hex encoded) values.
+function getNonUniqueKeyTypeValues (
+  maps:Map<Key, Value>|Map<Key, Value>[], 
+  keytype:KeyType
+) {
+  if (Array.isArray(maps)) {
+    // It's a set of input or output maps, so recursively check each map and set
+    // values.
+    const values:({ key:string, value:string|null})[][] = [];
+
+    for (const map of maps) {
+      values.push(
+        // There is probably a better way to type this
+        getNonUniqueKeyTypeValues(map, keytype) as { key:string, value:string|null }[]
+      );
+    }
+
+    return values;
+  }
+  
+  const map = maps;
+  const vals:{ key:string, value:string|null }[] = [];
+
+  for (const key of map.keys()) {
+    if (key.indexOf(keytype) === 0) {
+      const value = map.get(key)?.toString('hex') || null;
+      vals.push({ key, value });
+    }
+  }
+
+  return vals;
+}
+
+
+// A getter helper for optional keytypes which returns lists of values as hex
+// strings.
+function getOptionalMappedBytesAsHex (maps:Map<Key, Value>[], keytype:KeyType) {
+  const bytes:(string|null)[] = [];
+
+  for (const map of maps) {
+    bytes.push(map.get(keytype)?.toString('hex') ?? null);
+  }
+
+  return bytes;
+}
+
+
+// A getter helper for optional keytypes which returns lists of values as
+// numbers.
+function getOptionalMappedBytesAsUInt (maps:Map<Key, Value>[], keytype:KeyType) {
+  const bytes:(number|null)[] = [];
+
+  for (const map of maps) {
+    bytes.push(map.get(keytype)?.readUInt32LE() ?? null);
+  }
+
+  return bytes;
+}
+
+
+// Accepts a BIP0032 path as a string and returns a Buffer containing uint32
+// values for each path node.
+function parseDerivationPathNodesToBytes (path:string):Buffer {
+  const bw = new BufferWriter();
+
+  for (const node of path.match(BIP_32_NODE_REGEX) ?? []) {
+    // Skip slash and parse int
+    let num = parseInt(node.substring(1), 10);
+
+    if (node.indexOf('\'') > -1) {
+      // Hardened node needs hardening
+      num += BIP_32_HARDENING_OFFSET;
+    }
+
+    bw.writeU32(num);
+  }
+
+  return bw.render();
+}
```

Comment:
Just noting that we have another module in this library that might have some useful utilities for this, at the very least to reuse the regexes we already have. Checkout `paths.js`

Reply:
The `validateBIP32Path` is useful. The paths module doesn't export the regex values and they aren't as useful as what is here. The regex I use here lists nodes as match groups in a single operation. Additionally, the hardening offset is a constant instead of requiring an operation from `Math`.

I'd argue that it's probably the paths module which needs to be updated/improved with these features, but should it fit in the scope of the ledger-v2 feature?

Reply:
Agree yours is probably better.

As for scope, it's a tough question. You don't want to introduce scope creep but if you need an improvement to an existing feature to improve the feature within scope, and the scope creep isn't out of control, I would say it's probably worth it. We likely wouldn't have other opportunities to explicitly be fixing/updating something like this except within the scope of other issues. Similar also to increasing typescript coverage for example.